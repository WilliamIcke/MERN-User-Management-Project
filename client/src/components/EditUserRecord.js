import React from 'react'
import DatePicker from "react-datepicker";
import "react-datepicker/dist/react-datepicker.css";

const EditUserRecord = ({ userRecord, membershipRecords, updateUserRecord }) => {
    const [formData, setFormData] = React.useState(userRecord);
    const [disableForm, setDisableForm] = React.useState(false);
    const [validationMessage, setWordValidationMessage] = React.useState('');
    const [dateOfBirth, setDate] = React.useState(new Date(userRecord.dateOfBirth));

    React.useEffect(() => {
        // State changes, check valid data
        switch (true) {
            case (formData === undefined): // Check if value empty                    
                setDisableForm(true);
                break;
          case (formData.firstname.trim().length === 0): // Check if firstname is empty
                setDisableForm(true);
                setWordValidationMessage("Firstname is required");
                break;
          case (formData.surname.trim().length === 0): // Check if surname is empty
                  setDisableForm(true);
                  setWordValidationMessage("Surname is required");
                  break;
          case (formData.dateOfBirth === undefined || formData.dateOfBirth === null): // Check if DOB is set
                  setDisableForm(true);
                  setWordValidationMessage("Date of birth is required");
                break;
          default:
              setWordValidationMessage('');
              setDisableForm(false);
        }
    }, [formData, disableForm, dateOfBirth]);

    /* Updates form data with any input changes */
    const handleForm = (e) => {
        setFormData(Object.assign(Object.assign({}, formData), { [e.currentTarget.id]: e.currentTarget.value }));
    };

    /* Updates form data and datepicker field with date changes */
    const handleDateChange = (dateChange) => {
        setFormData(Object.assign(Object.assign({}, formData), { dateOfBirth: dateChange }));
        setDate(dateChange);
    };

    /* Updates membership changes */
    const handleMembershipChange = (e) => {
        setFormData(Object.assign(Object.assign({}, formData), { [e.currentTarget.id]: e.currentTarget.value }, { 'membershipStart': new Date() }));
    };

    return (
        <form className='EditForm' autoComplete="off" onSubmit={(e) => updateUserRecord(formData)}>
            <div className="form-container">
                <div className="row">
                    <div>
                        <label htmlFor='firstname'>Firstname</label>
                    </div>
                    <div>
                        <input onChange={handleForm} type='text' id='firstname' value={formData.firstname} />
                    </div>
                </div>
                <div className="row">
                    <div>
                        <label htmlFor='surname'>Surname</label>
                    </div>
                    <div>
                        <input onChange={handleForm} type='text' id='surname' value={formData.surname} />
                    </div>
                </div>
                <div className="row">
                    <div>
                        <label htmlFor='dateOfBirth'>Date of birth</label>
                    </div>
                    <div>
                        <DatePicker
                            selected={dateOfBirth}
                            onChange={handleDateChange}
                            id='dateOfBirth'
                            dateFormat='dd-MM-yyyy'
                            popperPlacement="bottom"
                        />
                    </div>
                </div>
                <div className="row">
                    <div>
                        <label htmlFor='membership'>Membership plan</label>
                    </div>
                    <div>
                        <select onChange={handleMembershipChange} id='membership' value={formData.membership._id}>
                            {membershipRecords.map((membershipRecord) => (
                                <option key={membershipRecord._id} value={membershipRecord._id}>
                                    {membershipRecord.membershipName}: Â£{membershipRecord.cost}
                                </option>
                            ))}
                        </select>
                    </div>
                </div>
                <p className="validationMessage"  data-testid="validation-message">{validationMessage}</p>
                <button className="userSubmit" disabled={disableForm} >Update User Record</button>
            </div>
        </form>
    )
}

export default EditUserRecord