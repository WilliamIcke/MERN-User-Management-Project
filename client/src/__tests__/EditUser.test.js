import React from 'react';
import { shallow, mount, render } from 'enzyme';
import EditUserRecord from '../components/EditUserRecord';

const testData = {
  userRecord: {
      _id: 1,
      firstname: 'Test',
      surname: 'Tester',
      dateOfBirth: '12/12/1990',
      membership: 1,
      membershipStart: '12/07/2018',
      active: true,
  },
  membershipRecords: [
      {
      _id: 1,
      membershipName: 'No Membership',
      cost: '0',
      active: true
      },
      {
      _id: 2,
      membershipName: 'Priority Membership',
      cost: '250',
      active: true
      },
      {
      _id: 3,
      membershipName: 'VIP Membership',
      cost: '500',
      active: true
      },
      {
      _id: 4,
      membershipName: 'Tier 1 Membership',
      cost: '120',
      active: true
      },
      {
      _id: 5,
      membershipName: 'Basic Membership',
      cost: '50',
      active: true
      }
  ]};


describe("Component shallow snapshot tests (debug)", () => {
  test('Check EditUserRecord component, only', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    expect(component).toMatchSnapshot();
  });
});

describe("Functionality elements", () => {
  test('Firstname field, delete all data, should result in update button being disabled and an appropriate validation message.', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#firstname');
    inputField.simulate('change', { currentTarget: { id:'firstname', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Firstname is required'));
    //expect(updateButton.disabled).toEqual('disabled');
  });

  test('Surname field, delete all data, should result in update button being disabled and an appropriate validation message', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#surname');
    inputField.simulate('change', { currentTarget: { id:'surname', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Surname is required'));
    //expect(updateButton.disabled).toEqual('disabled');
  });

  test('Date of birth field, delete all data, should result in update button being disabled and an appropriate validation message', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#dateOfBirth');
    inputField.simulate('change', { currentTarget: { id:'dateOfBirth', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Date of birth is required'));
    //expect(updateButton.disabled).toEqual('disabled');
  });

  test('Firstname field, make the field invalid then fix the field, should enabled the update button and remove validation message', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#firstname');
    inputField.simulate('change', { currentTarget: { id:'firstname', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Firstname is required'));
    //expect(updateButton.disabled).toEqual('disabled');
    
    inputField.simulate('change', { currentTarget: { id:'firstname', value: 'Test' } });
    
    expect(validationMessageElement.text().length == 0);
    //expect(updateButton.disabled).toEqual('');
  });

  test('Surname field, make the field invalid then fix the field, should enabled the update button and remove validation messag', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#surname');
    inputField.simulate('change', { currentTarget: { id:'surname', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Surname is required'));
    //expect(updateButton.disabled).toEqual('disabled');

    inputField.simulate('change', { currentTarget: { id:'surname', value: 'Tester' } });
    
    expect(validationMessageElement.text().length == 0);
    //expect(updateButton.disabled).toEqual('');
  });

  test('Date of birth field, make the field invalid then fix the field, should enabled the update button and remove validation messag', () => {
    const component =  shallow(
      <EditUserRecord 
        userRecord={testData.userRecord}
        membershipRecords={testData.membershipRecords} debug/>
    );
    const inputField = component.find('#dateOfBirth');
    inputField.simulate('change', { currentTarget: { id:'dateOfBirth', value: '' } });
    
    const validationMessageElement = component.find('#validation-message');
    const updateButton = component.find('#userSubmitButton');

    expect(validationMessageElement.text().includes('Date of birth is required'));
    //expect(updateButton.disabled).toEqual('disabled');

    inputField.simulate('change', { currentTarget: { id:'dateOfBirth', value: '12/12/2015' } });
    
    expect(validationMessageElement.text().length == 0);
    //expect(updateButton.disabled).toEqual('');
  });
});