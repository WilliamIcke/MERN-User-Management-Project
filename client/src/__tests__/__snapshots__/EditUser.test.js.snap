// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component shallow snapshot tests (debug) Check EditUserRecord component, only 1`] = `
<form
  autoComplete="off"
  className="EditForm"
  onSubmit={[Function]}
>
  <div
    className="form-container"
  >
    <div
      className="row"
    >
      <div>
        <label
          htmlFor="firstname"
        >
          Firstname
        </label>
      </div>
      <div>
        <input
          id="firstname"
          onChange={[Function]}
          type="text"
          value="Test"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div>
        <label
          htmlFor="surname"
        >
          Surname
        </label>
      </div>
      <div>
        <input
          id="surname"
          onChange={[Function]}
          type="text"
          value="Tester"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div>
        <label
          htmlFor="dateOfBirth"
        >
          Date of birth
        </label>
      </div>
      <div>
        <r
          allowSameDay={false}
          customTimeInput={null}
          dateFormat="dd-MM-yyyy"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="scroll"
          enableTabLoop={true}
          excludeScrollbar={true}
          focusSelectedMonth={false}
          id="dateOfBirth"
          monthsShown={1}
          nextMonthAriaLabel="Next Month"
          nextMonthButtonLabel="Next Month"
          nextYearAriaLabel="Next Year"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          popperPlacement="bottom"
          preventOpenOnFocus={false}
          previousMonthAriaLabel="Previous Month"
          previousMonthButtonLabel="Previous Month"
          previousYearAriaLabel="Previous Year"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={1990-12-12T00:00:00.000Z}
          selectsDisabledDaysInRange={false}
          shouldCloseOnSelect={true}
          showFourColumnMonthYearPicker={false}
          showFullMonthYearPicker={false}
          showMonthYearPicker={false}
          showPopperArrow={true}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showTwoColumnMonthYearPicker={false}
          showYearPicker={false}
          strictParsing={false}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          withPortal={false}
          yearItemNumber={12}
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div>
        <label
          htmlFor="membership"
        >
          Membership plan
        </label>
      </div>
      <div>
        <select
          id="membership"
          onChange={[Function]}
        >
          <option
            key="1"
            value={1}
          >
            No Membership
            : £
            0
          </option>
          <option
            key="2"
            value={2}
          >
            Priority Membership
            : £
            250
          </option>
          <option
            key="3"
            value={3}
          >
            VIP Membership
            : £
            500
          </option>
          <option
            key="4"
            value={4}
          >
            Tier 1 Membership
            : £
            120
          </option>
          <option
            key="5"
            value={5}
          >
            Basic Membership
            : £
            50
          </option>
        </select>
      </div>
    </div>
    <p
      className="validationMessage"
      id="validation-message"
    />
    <button
      className="userSubmit"
      disabled={false}
      id="userSubmitButton"
    >
      Update User Record
    </button>
  </div>
</form>
`;
